runtime/capsules/src/mctp/recv.rs

```diff
@@ -0,0 +1,232 @@
+/// on corresponding message_type.
+pub trait MCTPRxClient {
+    fn receive(&self, dst_eid: u8, msg_type: u8, msg_tag: u8, msg_payload: &[u8], msg_len: usize);
+}
+
+/// Receive state
+pub struct MCTPRxState<'a> {
+    /// Message assembly context
+    msg_terminus: MapCell<MsgTerminus>,
+    /// Expected message types
+    msg_types: Cell<&'static [MessageType]>,
+    /// Client (implements the MCTPRxClient trait)
+    client: OptionalCell<&'a dyn MCTPRxClient>,
+    /// Message buffer
+    msg_payload: TakeCell<'static, [u8]>,
+    /// next MCTPRxState node
+    next: ListLink<'a, MCTPRxState<'a>>,
+}
+
+impl<'a> ListNode<'a, MCTPRxState<'a>> for MCTPRxState<'a> {
+    fn next(&'a self) -> &'a ListLink<'a, MCTPRxState<'a>> {
+        &self.next
+    }
+}
+
+struct MsgTerminus {
+    msg_type: u8,
+    msg_tag: u8,
+    source_eid: u8,
+    tag_owner: u8,
+    start_payload_len: usize,
+    pkt_seq: u8,
+    msg_size: usize,
+}
+
+impl<'a> MCTPRxState<'a> {
+    pub fn new(
+        rx_msg_buf: &'static mut [u8],
+        message_types: &'static [MessageType],
+    ) -> MCTPRxState<'static> {
+        MCTPRxState {
+            msg_terminus: MapCell::empty(),
+            msg_types: Cell::new(message_types),
+            client: OptionalCell::empty(),
+            msg_payload: TakeCell::new(rx_msg_buf),
+            next: ListLink::empty(),
+        }
+    }
+
+    pub fn set_client(&self, client: &'a dyn MCTPRxClient) {
+        self.client.set(client);
+    }
+
+    /// Checks if a message of the given type is expected to be received.
+    ///
+    /// # Arguments
+    /// 'msg_type' - The message type to check if it is expected.
+    ///
+    /// # Returns
+    /// True if the message type is expected, false otherwise.
+    pub fn is_receive_expected(&self, msg_type: MessageType) -> bool {
+        let msg_types = self.msg_types.get();
+        for exp_msg_type in msg_types.iter() {
+            if msg_type == *exp_msg_type {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Checks from the received MCTP header if the next packet belongs to
+    /// the current message being assembled.
+    ///
+    /// # Arguments
+    /// 'mctp_hdr' - The MCTP header of the received packet.
+    /// 'pkt_payload_len' - The length of the payload of the received packet.
+    ///
+    /// # Returns
+    /// True if the next packet belongs to the current message, false otherwise.
+    pub fn is_next_packet(
+        &self,
+        mctp_hdr: &MCTPHeader<[u8; MCTP_HDR_SIZE]>,
+        pkt_payload_len: usize,
+    ) -> bool {
+        self.msg_terminus
+            .map(|msg_terminus| {
+                msg_terminus.tag_owner == mctp_hdr.tag_owner()
+                    && msg_terminus.msg_tag == mctp_hdr.msg_tag()
+                    && msg_terminus.source_eid == mctp_hdr.src_eid()
+                    && msg_terminus.pkt_seq == mctp_hdr.pkt_seq()
+                    && (mctp_hdr.som() == 0
```

Comment:
The next packet can never be the end packet, and must always be full length?

Reply:
This check is to ensure that all the middle packets (som == 0 and eom == 0) have the same length as the first packet. 
The full-length check can be skipped for last packet.

According to the following clause in MCTP base spec (section 8.4.2)
`The transmission unit of all packets in a given message shall be the same size, except for the transmission unit in the last packet (packet with EOM bit = 1b). `